packages/message-relayer/src/relay-tx.ts

```diff
@@ -0,0 +1,397 @@
+    stateBatchTransaction.data
+  )
+
+  return {
+    header: {
+      batchIndex: stateBatchAppendedEvent.args._batchIndex,
+      batchRoot: stateBatchAppendedEvent.args._batchRoot,
+      batchSize: stateBatchAppendedEvent.args._batchSize,
+      prevTotalElements: stateBatchAppendedEvent.args._prevTotalElements,
+      extraData: stateBatchAppendedEvent.args._extraData,
+    },
+    stateRoots,
+  }
+}
+
+/**
+ * Generates a Merkle proof (using the particular scheme we use within Lib_MerkleTree).
+ * @param leaves Leaves of the merkle tree.
+ * @param index Index to generate a proof for.
+ * @returns Merkle proof sibling leaves, as hex strings.
+ */
+const getMerkleTreeProof = (leaves: string[], index: number): string[] => {
+  const parsedLeaves = []
+  for (let i = 0; i < Math.pow(2, Math.ceil(Math.log2(leaves.length))); i++) {
+    if (i < leaves.length) {
+      parsedLeaves.push(leaves[i])
+    } else {
+      parsedLeaves.push(ethers.utils.keccak256('0x' + '00'.repeat(32)))
+    }
+  }
+
+  const bufLeaves = parsedLeaves.map(fromHexString)
+
+  const tree = new MerkleTree(
+    bufLeaves,
+    (el: Buffer | string): Buffer => {
+      return fromHexString(ethers.utils.keccak256(el))
+    }
+  )
+
+  const proof = tree.getProof(bufLeaves[index], index).map((element: any) => {
+    return toHexString(element.data)
+  })
+
+  return proof
+}
+
+/**
+ * Generates a Merkle-Patricia trie proof for a given account and storage slot.
+ * @param l2RpcProvider L2 RPC provider.
+ * @param blockNumber Block number to generate the proof at.
+ * @param address Address to generate the proof for.
+ * @param slot Storage slot to generate the proof for.
+ * @returns Account proof and storage proof.
+ */
+const getStateTrieProof = async (
+  l2RpcProvider: ethers.providers.JsonRpcProvider,
+  blockNumber: number,
+  address: string,
+  slot: string
+): Promise<StateTrieProof> => {
+  const proof = await l2RpcProvider.send('eth_getProof', [
+    address,
+    [slot],
+    toRpcHexString(blockNumber),
+  ])
+
+  return {
+    accountProof: toHexString(rlp.encode(proof.accountProof)),
+    storageProof: toHexString(rlp.encode(proof.storageProof[0].proof)),
+  }
+}
+
+/**
+ * Generates the transaction data to send to the L1CrossDomainMessenger in order to execute an
+ * L2 => L1 message.
+ * @param l1RpcProvider L1 RPC provider.
+ * @param l2RpcProvider L2 RPC provider.
+ * @param l1StateCommitmentChainAddress Address of the StateCommitmentChain.
+ * @param l2CrossDomainMessengerAddress Address of the L2CrossDomainMessenger.
+ * @param l2TransactionHash L2 transaction hash to generate a relay transaction for.
+ * @returns 0x-prefixed transaction data as a hex string.
+ */
+export const makeRelayTransactionData = async (
+  l1RpcProvider: ethers.providers.JsonRpcProvider,
+  l2RpcProvider: ethers.providers.JsonRpcProvider,
+  l1StateCommitmentChainAddress: string,
+  l2CrossDomainMessengerAddress: string,
+  l2TransactionHash: string
+): Promise<string> => {
```

Comment:
An alternative here is to return `Promise<{ message: CrossDomainMessage, proof: CrossDomainMessageProof }>`. Might be a bit more "tool friendly".

Reply:
It seems like a good idea

Reply:
Ok I've update the interface to be a lot nicer. Correctly handles multiple messages sent in the same transaction too.